### PR TITLE
refactor sysex code.

### PR DESCRIPTION
**Pull request by jmamma**
https://github.com/jmamma/MIDICtrl20_MegaCommand/issues/6

Work in progress.

### CODE REVIEW

**Comment by jmamma** [comment]
Most of the underlying DataEncoding code uses direct pointers to sysex data buffers.
This all needs to be refactored.

**Comment by jmamma** [comment]
I'm only updating the Decoder classes. The Encoder classes do not use Sysex buffer, and instead encode and write directly to the MIDI buffer.

*Correction. MNM library is using encoder to encode from packed-sysex to data

**Comment by jmamma** [comment]
@yatli This is working now.

I'm exhausted 😪 

If you have a chance please review/test the pull request. If satisfied I will merge in to DEV, then you need to merge/incorporate some of the minor A4 class changes to your new A4 code.

**Comment by yatli** [comment]
Great work @jmamma! And get some good rest! 
Let me first go through the code and make minor touches if necessary. When I get back home tomorrow, I'll do a test drive.

File: avr/cores/megacommand/CommonTools/DataEncoderUnchecking.hh [hunk 1]
```diff
@@ -21,7 +21,11 @@
 #define DATA_ENCODER_TRUE()  return
 #define DATA_ENCODER_FALSE() return
 
-#define DATA_ENCODER_INIT(data, length) data
+//Required for Macro argument overloading
+#define GET_MACRO(_1,_2,_3,NAME,...) NAME
```
Comment (yatli) [inline avr/cores/megacommand/CommonTools/DataEncoderUnchecking.hh:25]: learned something today!

**Comment by jmamma** [comment]
This pull request appears stable.

**Comment by yatli** [comment]
Sorry I'm delayed a bit; The schedule is a bit tight for my day job.

**Review by yatli — APPROVED** [review]
LGTM. Mostly transformations from raw buffer to `MidiSysex`;
I've got some idea to fold the `if (ptr) { ... } else {...}` branches, but I can work on it in another PR.

File: avr/cores/megacommand/Midi/MidiSysex.hh [hunk 2]
```diff
@@ -69,18 +69,50 @@ protected:
...
   void resetRecord(uint8_t *buf = NULL, uint16_t maxLen = 0);
-  bool recordByte(uint8_t c);
+
+  bool putByte(uint16_t offset, uint8_t c) {
+    put_byte_bank1(sysex_highmem_buf + offset, c);
+  }
+
+  uint8_t getByte(uint16_t n) {
+    if (n < maxRecordLen) {
+      // Record data to specified memory buffer
```
Comment (yatli) [inline avr/cores/megacommand/Midi/MidiSysex.hh]: nit: Record -> Retrieve

File: avr/cores/megacommand/Midi/MidiSysex.hh [hunk 3]
```diff
@@ -69,18 +69,50 @@ protected:
...
+      if (recordBuf != NULL) {
+        return recordBuf[n];
+
+      } else {
+       // Write to sysex buffers in HIGH membank
```
Comment (yatli) [inline avr/cores/megacommand/Midi/MidiSysex.hh]: nit: Write -> Read

File: avr/cores/megacommand/Midi/MidiSysex.hh [hunk 4]
```diff
@@ -69,18 +69,50 @@ protected:
   bool recording;
   uint8_t recvIds[3];
   bool sysexLongId;
+  uint8_t *data;
+  uint8_t *recordBuf;
+  volatile uint8_t *sysex_highmem_buf;
 
 public:
   void startRecord(uint8_t *buf = NULL, uint16_t maxLen = 0);
```
Comment (yatli) [inline avr/cores/megacommand/Midi/MidiSysex.hh:77]: I see `maxLen` is stored to `maxRecordLen` no matter it's a supplied buf or `sysex_highmem_buf` -- we then have a few `if (n < maxRecordLen)` checks..

**Comment by yatli** [comment]
I'm playing with this branch. Looks stable to me but I guess that's because I'm only scratching the surface  :p

File: avr/cores/megacommand/Midi/MidiSysex.hh [hunk 5]
```diff
@@ -79,18 +79,18 @@ public:
 
   void resetRecord(uint8_t *buf = NULL, uint16_t maxLen = 0);
 
-  bool putByte(uint16_t offset, uint8_t c) {
+  inline void putByte(uint16_t offset, uint8_t c) {
```
Comment (jmamma) [inline avr/cores/megacommand/Midi/MidiSysex.hh]: "Any function defined right in the class definition is implicitly marked inline."

My assumption here. Do you agree with this statement?

Comment (yatli) [reply avr/cores/megacommand/Midi/MidiSysex.hh]: According to the spec: `A function defined entirely inside a class/struct/union definition, whether it's a member function or a non-member friend function, is implicitly an inline function.`
So, yes :) I'm reverting this bit.

Comment (jmamma) [reply avr/cores/megacommand/Midi/MidiSysex.hh]: Good. I was only made aware of this one recently. Some header files I have explicitly used "inline" in the past. Will need to clean this up.

File: avr/cores/megacommand/MCL/GridPage.cpp [hunk 6]
```diff
@@ -622,17 +622,18 @@ void GridPage::apply_slot_changes() {
...
     if (slot_clear == 1) {
       grid.clear_slot(track + getCol(), getRow());
       reload_slot_models = true;
     } else if (slot_copy == 1) {
       SET_BIT32(slot_buffer_mask, track + getCol());
+    } else {
+      slot.active = row_headers[cur_row].track_type[track + getCol()];
+      //  if (slot.active != EMPTY_TRACK_TYPE) {
+      slot.store_track_in_grid(track + getCol(), getRow());
+      proj.file.sync();
```
Comment (jmamma) [inline avr/cores/megacommand/MCL/GridPage.cpp]: proj.file.sync() should be outside bracket. or also included in if (slot_clear == 1)

Comment (yatli) [reply avr/cores/megacommand/MCL/GridPage.cpp]: done

**Pull request merged**